File: src/main/scala/fungsi/src/Lexer.scala
package fungsi.src

import scala.annotation.tailrec

object Lexer:
  def scan(code: String): List[Token] =
    val symbols = "+-\\|/?.><!#@`^~%&*-_+="
    val keywords = Map("def" -> Token.DEF)
  
    @tailrec
    def help(tokens: List[Token], code: String): List[Token] =
      if code == "" then
        return (new Token(Token.EOF, "") +: tokens).reverse
      val c = code.head
      c match
        case '@' => help(new Token(Token.LAMBDA, "@") +: tokens, code.tail)
        case '$' => help(new Token(Token.QUOTE, "~") +: tokens, code.tail)
        case '^' => help(new Token(Token.RETURN, "^") +: tokens, code.tail)
        case '|' => help(new Token(Token.DELIMITER, "|") +: tokens, code.tail)
        case ';' => help(new Token(Token.SEMICOLON, ";") +: tokens, code.tail)
        case ',' => help(new Token(Token.COMMA, ",") +: tokens, code.tail)
        case ':' =>
          if code.tail.head == '=' then
            help(new Token(Token.BIND, ":=") +: tokens, code.tail.tail)
          else
            val name = addName(code.tail, "")
            help(new Token(Token.NAME, ":" + name._1) +: tokens, name._2)
        case '(' | '[' => help(new Token(Token.LPAREN, "(") +: tokens, code.tail)
        case ')' | ']' => help(new Token(Token.RPAREN, ")") +: tokens, code.tail)
        case '\r' | '\t' | ' ' => help(tokens, code.tail)
        case '\'' =>
          val seq = addSeq(code.tail, "", '\'')
          help(new Token(Token.CHAR, seq._1) +: tokens, seq._2)
        case '"' =>
          val seq = addSeq(code.tail, "", '"')
          help(new Token(Token.STRING, seq._1) +: tokens, seq._2)
        case _ if c.isDigit =>
          val num = addNum(code, "")
          help(new Token(Token.NUM, num._1) +: tokens, num._2)
        case _ if c.isLetter || symbols.contains(c) =>
          val name = addName(code, "")
          keywords get name._1 match {
            case Some(tok) if tok == Token.DEF =>
              help(new Token(keywords(name._1), name._1) +: tokens, name._2)
            case None | Some(_) =>
              help(new Token(Token.NAME, name._1) +: tokens, name._2)
          }
        case _ =>
          println("lexer error")
          Nil
    end help
  
    @tailrec
    def addSeq(code: String, res: String, end: Char): (String, String) =
      if code.head == end then
        return (res.reverse, code.tail)
      addSeq(code.tail, code.head +: res, end)
    end addSeq
  
    @tailrec
    def addNum(code: String, res: String): (String, String) =
      def isDigit(c: Char): Boolean = c.isDigit || c == '.'
  
      if !isDigit(code.head) then
        return (res.reverse, code)
      addNum(code.tail, code.head +: res)
    end addNum
    
    @tailrec
    def addName(code: String, res: String): (String, String) =
      def isLetter(c: Char): Boolean = c.isLetter || c.isDigit || symbols.contains(c)
  
      if !isLetter(code.head) then
        return (res.reverse.toLowerCase, code)
      addName(code.tail, code.head +: res)
    end addName
    
    help(Nil, code)
  end scan
end Lexer

File: depr/haskell/src/Lexer.hs
{-# OPTIONS_GHC -Wno-incomplete-patterns #-}

module Lexer where

import Data.Char
import Debug.Trace
import Help
import Token

scan :: String -> [Token]
scan [] = []
scan source = HELP [] source
  where
    HELP :: [Token] -> String -> [Token]
    HELP tokens "" = reverse $ Token EOF "eof" None : tokens
    HELP tokens (':' : '=' : ode) = HELP (Token BIND ":=" InfixOp : tokens) ode
    HELP tokens code@(c : ode) =
      case c of
        '+' -> HELP (Token PLUS "+" InfixOp : tokens) ode
        '*' -> HELP (Token STAR "*" InfixOp : tokens) ode
        '|' -> HELP (Token VBAR "|" InfixOp : tokens) ode
        ',' -> HELP (Token COMMA "," InfixOp : tokens) ode
        '\n' -> HELP (simpleToken SEMICOLON "\n" : tokens) ode
        '@' -> HELP (simpleToken LAMBDA "->" : tokens) ode
        '-' -> HELP (Token MINUS "-" InfixOp : tokens) ode
        _
          | c == '/' || c == '\\' || c == '!' ->
            if head ode == '='
              then HELP (Token NEQUAL "/=" InfixOp : tokens) $ slice ode 1 $ length ode
              else HELP (Token SLASH "/" InfixOp : tokens) ode
        _ | c == '(' || c == '[' -> HELP (simpleToken LPAREN "(" : tokens) ode
        _ | c == ')' || c == ']' -> HELP (simpleToken RPAREN ")" : tokens) ode
        _ | c == '=' || c == '<' || c == '>' -> addEqual code
        _ | isNumber c -> addNum code False ""
        _ | isLetter c -> addName code
        _ | c == ' ' || c == '\t' || c == '\r' -> HELP tokens ode
        _ | otherwise -> error $ "Syntax error: " ++ show c
      where
        simpleToken :: TokenType -> String -> Token
        simpleToken token content = Token token content None

        addEqual :: String -> [Token]
        addEqual ('<' : '=' : xs) = HELP (Token LEQUAL "<=" InfixOp : tokens) xs
        addEqual ('>' : '=' : xs) = HELP (Token GEQUAL ">=" InfixOp : tokens) xs
        addEqual ('=' : '=' : xs) = HELP (Token DEQUAL "==" InfixOp : tokens) xs
        addEqual ('<' : xs) = HELP (Token LESS "<" InfixOp : tokens) ode
        addEqual ('>' : xs) = HELP (Token GREAT ">" InfixOp : tokens) ode
        addEqual ('=' : xs) = HELP (Token EQUAL "=" InfixOp : tokens) ode

        addNum :: String -> Bool -> String -> [Token]
        addNum ('.' : xs) isFloat res = addNum xs True (res ++ ".")
        addNum xs@(x : s) isFloat res
          | isNumber x = addNum s isFloat (res ++ [x])
          | otherwise = resultNum isFloat res xs
        addNum [] isFloat res = resultNum isFloat res []

        resultNum :: Bool -> String -> String -> [Token]
        resultNum True res next = HELP (Token FLOAT res Num : tokens) next
        resultNum False res next = HELP (Token INT res Num : tokens) next

        addName code = HELP (Token (fst4 token) (snd4 token) (thd4 token) : tokens) (fth4 token)
          where
            token = case code of
              ('i' : 'f' : xs) -> (IF, "if", Name, xs)
              ('t' : 'h' : 'e' : 'n' : xs) -> (THEN, "then", None, xs)
              ('e' : 'l' : 's' : 'e' : xs) -> (ELSE, "else", None, xs)
              ('t' : 'r' : 'u' : 'e' : xs) -> (TRUE, "true", Name, xs)
              ('f' : 'a' : 'l' : 's' : 'e' : xs) -> (FALSE, "false", Name, xs)
              ('a' : 'n' : 'd' : xs) -> (AND, "and", Name, xs)
              ('o' : 'r' : xs) -> (OR, "or", Name, xs)
              ('n' : 'o' : 't' : xs) -> (NOT, "not", Name, xs)
              ('n' : 'i' : 'l' : xs) -> (NIL, "nil", Name, xs)
              xs ->
                ( NAME,
                  takeWhile isName code,
                  Name,
                  dropWhile isName xs
                )

isName :: Char -> Bool
isName c = isLetter c || isNumber c